middleware.js

```diff
@@ -0,0 +1,71 @@
+import { NextResponse } from "next/server";
+import { Buffer } from "node:buffer";
+
+const timeoutController = new Promise((resolve) => {
+  setTimeout(resolve, 150);
+});
+
+async function sendMetric(hostName, metricKey) {
+  fetch(`${hostName}/api/metric`, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json"
+    },
+    body: JSON.stringify({ metricKey }),
+  });
+}
+
+export async function middleware(req) {
+  const cookie = req.headers.get("Cookie");
+  const split_url = req.url.split("/");
+  const HOST_NAME = split_url.slice(0, 3).join("/");
+  let _metricName;
+  if (req.url.includes("profile")) {
+    _metricName = split_url.slice(3, -1).join("_");
+  } else _metricName = split_url.slice(3).join("_");
+
+  // console.log(HOST_NAME, _metricName);
+
+  // skip the /api/metric api endpoint
+  if (req.url.includes("metric")) return NextResponse.json({ success: true });
+
+  try {
+    let response;
+    if (req.body) {
+      const rawBody = await req.body.getReader().read();
+      const body = JSON.parse(Buffer.from(rawBody?.value).toString("utf8"));
+      // const reqHeaders = Object.fromEntries(req.headers.entries());
+
+      response = await fetch(req.url, {
+        method: req.method,
+        headers: {
+          "Content-Type": "application/json",
+          "Cookie": cookie
+        },
+        body: JSON.stringify(body)
+      });
+
+    } else {
+      response = await fetch(req.url, {
+        headers: {
+          "Cookie": cookie
+        }
+      });
+    }
+
+    const data = await response.json();
+
+    // send metric and ignore failure
```

Comment:
Suggest making this comment a bit more accurate - attempt to send metric, but early-out after 150ms.  Additionally (nitpick!) would suggest making the timeoutController created with a function (createTimeoutPromise?) that takes in the amount of time as a parameter - this would make the line nearly self-documented